torch/distributed/_sharded_optim/api.py

```diff
@@ -0,0 +1,61 @@
+from typing import List, Union, Mapping, Dict, Any
+
+from torch import Tensor
+from torch.distributed._sharded_tensor import ShardedTensor
+
+
+class ShardedOptimizer(object):
+    def __init__(
+        self,
+        params: Mapping[str, Union[Tensor, ShardedTensor]],
+        optimizer_class,
+        *optimizer_args,
+        **optimizer_kwargs
+    ):
+        """
+        Collects all Tensors and local shards of ShardedTensor, uses these
+        Tensors as ``params`` for the optimizer while building the optimizer
+        using ``optimizer_class``, ``*optimizer_args`` and
+        ``*optimizer_kwargs``.
+        """
+        tensors: List[Tensor] = []
+        for value in params.values():
+            if isinstance(value, ShardedTensor):
+                for local_shard in value.local_shards():
+                    tensors.append(local_shard.tensor)
+            else:
+                tensors.append(value)
+
+        self.params = params
+        self._optim = optimizer_class(tensors, *optimizer_args, **optimizer_kwargs)
+        self.param_groups = self._optim.param_groups
+        self.state = self._optim.state
+
+    def zero_grad(self, set_to_none: bool = False):
+        self._optim.zero_grad(set_to_none)
+
+    def step(self, closure=None):
+        """
+        Regular optimizer step
+        """
+        self._optim.step(closure)
+
+    def state_dict(self) -> Dict[str, Any]:
+        """
+        Returned state and param_groups will contain parameter keys
+        instead of parameter indices in torch.Optimizer.
```

Comment:
Do you think it makes sense to have a "tensor" only interface like `Dict[str, Tensor]` in `ShardedOptimizer`?
If the vanilla optimizer use primitives, we can do the wrapping here.

Reply:
@zzzwen do you mean the "tensor" only interface for `ShardedOptimizer.state_dict()`? or you are referring to sth else

Reply:
Yes, I am talking about `ShardedOptimizer.state_dict()`

Reply:
I see, yeah `state_dict` it might make sense to have a tensor only interface, it will be done in a follow up PR https://github.com/pytorch/pytorch/pull/68608, still thinking about the correct semantics, lets discuss more there.

Reply:
@zzzwen Not sure I understood the proposal here. Are you proposing adding two state_dict methods, one that returns Dict[str, Any] and one that returns Dict[str, Tensor]?